File: java/src/pattern/strategy/duck/Duck.java
package pattern.strategy.duck;

public abstract class Duck
{
    protected FlyBehaviour flyBehaviour;
    protected QuackBehaviour quackBehaviour;

    public void Duck()
    {
    }

    public void quack()
    {
	quackBehaviour.quack();
    }

    public void fly()
    {
	flyBehaviour.fly();
    }

    public abstract void display();

    public void swim()
    {
	System.out.println("All ducks float, even decoys!");
    }

    public void setFlyBehaviour(FlyBehaviour flyBehaviour)
    {
	this.flyBehaviour = flyBehaviour;
    }

    public void setQuackBehaviour(QuackBehaviour quackBehaviour)
    {
	this.quackBehaviour = quackBehaviour;
    }
}

File: php/src/pattern/strategy/duck/Duck.php
<?php

namespace pattern\strategy\duck;

abstract class Duck
{
    protected $_flyBehaviour;
    protected $_quackBehaviour;

    public function __construct()
    {
    }

    public function quack()
    {
	$this->_quackBehaviour->quack();
    }

    public function fly()
    {
	$this->_flyBehaviour->fly();
    }

    public abstract function display();

    public function swim()
    {
	echo "All ducks float, even decoys!\n";
    }

    public function setFlyBehaviour(FlyBehaviour $flyBehaviour)
    {
	$this->_flyBehaviour = $flyBehaviour;
    }

    public function setQuackBehaviour(QuackBehaviour $quackBehaviour)
    {
	$this->_quackBehaviour = $quackBehaviour;
    }
}
